File: core/tests/layer2/test_shockley_diode.cpp
// =============================================================================
// Pulsim — ShockleyDiode (exponential junction), Phase 4 audit C.1
// =============================================================================

#include "pulsim/ad/ad_scalar.hpp"
#include "pulsim/models/device_model.hpp"
#include "pulsim/models/shockley_diode.hpp"

#include <catch2/catch_approx.hpp>
#include <catch2/catch_test_macros.hpp>

#include <array>
#include <cmath>

using namespace pulsim;
using namespace pulsim::models;
using Catch::Approx;

namespace {

Real iD(const ShockleyDiode::Params& p, Real v) {
    const Real vv[2] = {v, 0.0};
    return ShockleyDiode::current<Real>(vv, p);
}

}  // namespace

TEST_CASE("ShockleyDiode — matches the closed form over decades",
          "[v2][layer2][shockley_diode][unit]") {
    ShockleyDiode::Params p;   // I_S = 1e-12, n = 1, V_T = 25.852 mV
    for (const Real i : {1e-6, 1e-3, 1.0, 10.0, 1e3}) {
        const Real v = ShockleyDiode::voltage_for_current(p, i);
        // `voltage_for_current` inverts the JUNCTION term only,
        // so the law's own G_min leakage rides on top. At 1 µA
        // that is 3.6e-13 A out of 1e-6 — a 3.6e-7 relative
        // offset, which is the model working, not drifting.
        INFO("i = " << i << " A, v = " << v << " V");
        REQUIRE(iD(p, v) == Approx(i + v * p.G_min)
                               .epsilon(1e-12));
    }
}

TEST_CASE("ShockleyDiode — V_F rises ~60 mV per decade",
          "[v2][layer2][shockley_diode][unit]") {
    ShockleyDiode::Params p;
    const Real v1 = ShockleyDiode::voltage_for_current(p, 1e-3);
    const Real v2 = ShockleyDiode::voltage_for_current(p, 1e-2);
    // n·V_T·ln(10) = 59.5 mV for n = 1.
    REQUIRE(v2 - v1 == Approx(p.V_T * std::log(10.0))
                          .epsilon(1e-6));
}

TEST_CASE("ShockleyDiode — the working range is NOT a tangent",
          "[v2][layer2][shockley_diode][unit][regression]") {
    // The mistake this model was first written with. Continuing
    // the exponential by its tangent above SPICE's `vcrit`
    // (≈ 0.63 V, i.e. ≈ 18 mA) turns the device into a 1.41 Ω
    // resistor: 14.7 V at 10 A instead of 0.77 V. It converged
    // happily while doing it, which is why this is pinned.
    ShockleyDiode::Params p;
    for (const Real i : {0.018, 1.0, 10.0, 1e3, 1e5}) {
        const Real v = ShockleyDiode::voltage_for_current(p, i);
        INFO("i = " << i << " A");
        REQUIRE(v < 1.1);                     // a junction
        REQUIRE(iD(p, v) == Approx(i).epsilon(1e-9));
    }
}

TEST_CASE("ShockleyDiode — finite for voltages Newton invents",
          "[v2][layer2][shockley_diode][unit]") {
    // exp(50/0.02585) = e^1934 is +inf. The limiter's whole job.
    ShockleyDiode::Params p;
    for (const Real v : {5.0, 50.0, 500.0, 5000.0}) {
        const Real i = iD(p, v);
        INFO("v = " << v << " V");
        REQUIRE(std::isfinite(i));
        REQUIRE(i > 0.0);
    }
    // Monotone all the way out, so Newton always has a descent
    // direction back toward the solution.
    REQUIRE(iD(p, 500.0) > iD(p, 50.0));
    REQUIRE(iD(p, 50.0) > iD(p, 5.0));
}

TEST_CASE("ShockleyDiode — Jacobian is finite and matches FD",
          "[v2][layer2][shockley_diode][ad]") {
    ShockleyDiode::Params p;
    for (const Real v : {-100.0, -1.0, 0.0, 0.5, 0.7, 50.0}) {
        const std::array<Real, 2> vv{v, 0.0};
        const auto [i, d] =
            evaluate_current_and_jacobian<ShockleyDiode>(vv, p);
        INFO("v = " << v);
        REQUIRE(std::isfinite(i));
        REQUIRE(std::isfinite(d[0]));
        REQUIRE(std::isfinite(d[1]));
        // dI/dv_anode must be positive: the junction is monotone.
        REQUIRE(d[0] > 0.0);
        // Two terminals, one branch current: the partials are
        // equal and opposite.
        REQUIRE(d[1] == Approx(-d[0]).epsilon(1e-12));
    }
}

TEST_CASE("ShockleyDiode — C1 across the limiter join",
          "[v2][layer2][shockley_diode][ad]") {
    ShockleyDiode::Params p;
    const Real vl = ShockleyDiode::v_lim(p);
    const Real h  = 1e-7;
    const std::array<Real, 2> below{vl - h, 0.0};
    const std::array<Real, 2> above{vl + h, 0.0};
    const auto [i_b, d_b] =
        evaluate_current_and_jacobian<ShockleyDiode>(below, p);
    const auto [i_a, d_a] =
        evaluate_current_and_jacobian<ShockleyDiode>(above, p);

    // Both assertions below are bounded by the curve's own
    // variation over the probe distance, not by a fixed epsilon.
    // A fixed epsilon here would be asserting that the function
    // is FLAT near the join, which it is not and should not be —
    // continuity is the claim, and the probes sit h away on
    // either side of it.
    //
    // SLOPE: above the join the derivative is the constant g_l;
    // below it the exponential's own derivative, smaller by
    // exp(-h/vte). So they differ by ~2h/vte = 7.7e-6 relative,
    // and anything at that scale is the exponential's curvature,
    // not a kink.
    const Real vte = p.n * p.V_T;
    REQUIRE(std::abs(d_a[0] - d_b[0])
            <= d_a[0] * (2.0 * h / vte) * 1.01);

    // VALUE: g ≈ 3.9e7 S here, so 2h = 2e-7 V of separation is
    // 7.7 A of legitimate difference on top of 1e6 A.
    REQUIRE(std::abs(i_a - i_b) <= 2.0 * h * d_a[0] * 1.001);

    // And the join really is a join: extrapolating the lower
    // branch's tangent across it lands on the upper branch to
    // full double precision, which a C0-but-not-C1 model could
    // not do.
    REQUIRE(i_b + d_b[0] * (2.0 * h) ==
              Approx(i_a).epsilon(1e-10));
}

TEST_CASE("ShockleyDiode — blocks in reverse, breaks down on demand",
          "[v2][layer2][shockley_diode][unit]") {
    ShockleyDiode::Params p;
    REQUIRE(std::abs(iD(p, -100.0)) < 1e-9);   // I_S + G_min only

    ShockleyDiode::Params z = p;
    z.BV = 5.1;
    REQUIRE(std::abs(iD(z, -4.0)) < 1e-9);     // still blocking
    REQUIRE(iD(z, -5.5) < -1e-6);              // conducting back
    REQUIRE(std::isfinite(iD(z, -500.0)));     // and still finite
}

TEST_CASE("ShockleyDiode — the tempco needs I_S, not just kT/q",
          "[v2][layer2][shockley_diode][unit]") {
    // Raising V_T alone RAISES the drop — the opposite of the
    // familiar behaviour. The real negative tempco comes from
    // I_S, which roughly doubles every 10 °C.
    constexpr Real T_cold = 298.15, T_hot = 398.15;
    ShockleyDiode::Params cold;
    cold.V_T = ShockleyDiode::thermal_voltage(T_cold);
    ShockleyDiode::Params vt_only = cold;
    vt_only.V_T = ShockleyDiode::thermal_voltage(T_hot);

    const Real v_cold =
        ShockleyDiode::voltage_for_current(cold, 1.0);
    const Real v_vt_only =
        ShockleyDiode::voltage_for_current(vt_only, 1.0);
    REQUIRE(v_vt_only > v_cold);

    ShockleyDiode::Params hot = vt_only;
    hot.I_S = ShockleyDiode::saturation_current_at(
        cold.I_S, T_hot, T_cold);
    const Real v_hot =
        ShockleyDiode::voltage_for_current(hot, 1.0);
    REQUIRE(v_hot < v_cold);
    const Real tempco = (v_hot - v_cold) / (T_hot - T_cold);
    INFO("tempco = " << tempco << " V/K");
    REQUIRE(tempco < -1e-3);
    REQUIRE(tempco > -3e-3);
}

TEST_CASE("ShockleyDiode — unphysical parameters are refused",
          "[v2][layer2][shockley_diode][validation]") {
    ShockleyDiode::Params p;
    REQUIRE_NOTHROW(ShockleyDiode::validate(p));

    auto bad = [](auto mutate) {
        ShockleyDiode::Params q;
        mutate(q);
        REQUIRE_THROWS_AS(ShockleyDiode::validate(q),
                           std::invalid_argument);
    };
    bad([](auto& q) { q.I_S = 0.0; });
    bad([](auto& q) { q.n = 0.0; });
    bad([](auto& q) { q.V_T = -1.0; });
    bad([](auto& q) { q.G_min = -1.0; });
    bad([](auto& q) { q.i_lim = 0.0; });
    bad([](auto& q) { q.BV = -5.1; });
}

File: python/tests/test_shockley_diode.py
"""The exponential (Shockley) diode (audit C.1).

Pulsim had two diodes and neither was exponential: `add_diode` is
binary PWL, and the smooth-blend nonlinear diode is a sigmoid
blended onto a STRAIGHT LINE, i = (v - V_F0)/R_d above the knee.
Both fix the forward drop by construction.

A real junction does not. V_F rises about 60 mV per decade of
current, so the same device that drops 0.53 V at 1 mA drops
0.77 V at 10 A. Over the decades a converter spans between light
and full load, a fixed-drop model is wrong by a couple of hundred
millivolts in the quantity that sets conduction loss -- and wrong
in OPPOSITE directions at the two ends, which is an error no
single fitted V_F0 can remove.
"""

import math

import numpy as np
import pytest

import pulsim as p

I_S, N, V_T = 1e-12, 1.0, 0.025852


def _v_forward(current, **kw):
    """Forward voltage carried by a forced current."""
    b = p.CircuitBuilder()
    # add_current_source(from, to) DRAINS `to`, so this pushes
    # `current` INTO node "a".
    b.add_current_source("I1", "a", "gnd", current)
    b.add_shockley_diode("D1", "a", "gnd", **kw)
    res = p.simulate(b, t_end=1e-9, dt=1e-9, engine="pwl")
    return float(np.asarray(res.v("a"))[-1])


def _analytic(i, I_S=I_S, n=N, V_T=V_T):
    return n * V_T * math.log(i / I_S + 1.0)


def test_it_follows_the_shockley_law_over_seven_decades():
    for i in (1e-6, 1e-4, 1e-2, 1.0, 10.0, 100.0):
        assert _v_forward(i) == pytest.approx(_analytic(i),
                                               rel=1e-4), i


def test_v_f_rises_sixty_millivolts_per_decade():
    """The property the fixed-drop models cannot have. For n = 1
    the slope is n*V_T*ln(10) = 59.5 mV/decade."""
    v1, v2 = _v_forward(1e-3), _v_forward(1e-2)
    assert (v2 - v1) == pytest.approx(V_T * math.log(10.0),
                                       rel=1e-3)


def test_the_working_range_is_the_exponential_not_a_tangent():
    """The mistake this model was FIRST written with, kept as a
    test because it is silent and plausible.

    The exponent has to be limited or Newton's first trial step
    (v ~ 50 V, e^1934) is +inf. SPICE does that with `pnjlim`,
    which limits the per-iteration STEP -- the converged answer
    still sits on the true curve. Continuing the curve by its own
    tangent above SPICE's `vcrit` instead LOOKS equivalent and is
    not: vcrit is only ~0.63 V, i.e. ~18 mA, and the tangent from
    there is a 1.41 ohm resistor. That model reported 14.7 V at
    10 A instead of 0.77 V, and converged happily while doing it.

    So: pin the physical range against the closed form, hard.
    """
    for i in (0.018, 0.1, 1.0, 10.0, 100.0, 1000.0):
        v = _v_forward(i)
        assert v == pytest.approx(_analytic(i), rel=1e-4), i
        assert v < 1.2, (i, v)      # a junction, not a resistor


def test_the_limiter_only_lives_above_any_real_current():
    """Where the tangent DOES take over, and that it is far away.
    Default i_lim = 1e6 A -- four orders past the largest
    converter anyone simulates."""
    v_lim = _analytic(1e6)
    assert 1.0 < v_lim < 1.15, v_lim
    # Just below it the law is still exactly exponential.
    assert _v_forward(9e5) == pytest.approx(_analytic(9e5),
                                             rel=1e-4)


def test_it_survives_a_voltage_newton_would_never_reach():
    """The reason the limiter exists at all: without it,
    exp(50/0.02585) = e^1934 is +inf, the Jacobian is +inf, the
    update is NaN and the run is over. A stiff forward-biased
    source must simply converge."""
    b = p.CircuitBuilder()
    b.add_voltage_source("V1", "s", "gnd", 50.0)
    b.add_resistor("R1", "s", "a", 5.0)
    b.add_shockley_diode("D1", "a", "gnd")
    res = p.simulate(b, t_end=1e-8, dt=1e-9, engine="pwl")
    v = float(np.asarray(res.v("a"))[-1])
    i = (50.0 - v) / 5.0
    assert v == pytest.approx(_analytic(i), rel=1e-3), (v, i)


def test_it_blocks_in_reverse():
    b = p.CircuitBuilder()
    b.add_voltage_source("V1", "a", "gnd", -100.0)
    b.add_shockley_diode("D1", "a", "gnd")
    res = p.simulate(b, t_end=1e-8, dt=1e-9, engine="pwl")
    i = abs(float(np.asarray(res.i("V1"))[-1]))
    assert i < 1e-6, i          # I_S plus G_min leakage only


def test_breakdown_conducts_past_the_knee_and_not_before():
    """A 5.1 V Zener: blocking at 4 V, conducting at 6 V."""
    def rev(v_rev):
        b = p.CircuitBuilder()
        b.add_voltage_source("V1", "s", "gnd", -v_rev)
        b.add_resistor("R1", "s", "a", 10.0)
        b.add_shockley_diode("D1", "a", "gnd", BV=5.1)
        res = p.simulate(b, t_end=1e-8, dt=1e-9, engine="pwl")
        return float(np.asarray(res.v("a"))[-1])

    assert rev(4.0) == pytest.approx(-4.0, abs=0.01)   # blocking
    v6 = rev(6.0)
    assert -6.0 < v6 < -5.1, v6                        # clamped


def test_raising_only_v_t_raises_the_drop():
    """The opposite of what everyone remembers, so it is pinned.

    At fixed current and fixed I_S, V_F = n*V_T*ln(i/I_S), so V_T
    and V_F move TOGETHER. Sweeping temperature through
    `thermal_voltage` alone therefore makes the diode drop MORE,
    not less. (An earlier version of this test asserted the
    familiar negative tempco and failed at 0.948 V vs 0.710 V --
    the expectation was wrong, and so was the docstring that had
    promised it.)
    """
    v_cold = _v_forward(1.0, V_T=p.thermal_voltage(298.15))
    v_hot_vt_only = _v_forward(1.0, V_T=p.thermal_voltage(398.15))
    assert v_hot_vt_only > v_cold
    assert v_hot_vt_only / v_cold == pytest.approx(
        398.15 / 298.15, rel=1e-3)


def test_the_negative_tempco_needs_i_s_as_well():
    """Where the real -2 mV/K comes from: I_S roughly doubles
    every 10 C, and that term overwhelms kT/q. Scale both and the
    familiar behaviour appears."""
    T_cold, T_hot = 298.15, 398.15
    v_cold = _v_forward(1.0, V_T=p.thermal_voltage(T_cold),
                        I_S=p.shockley_saturation_current_at(
                            I_S, T_cold))
    v_hot = _v_forward(1.0, V_T=p.thermal_voltage(T_hot),
                       I_S=p.shockley_saturation_current_at(
                           I_S, T_hot))
    assert v_hot < v_cold, (v_cold, v_hot)
    tempco = (v_hot - v_cold) / (T_hot - T_cold)
    assert -3e-3 < tempco < -1e-3, tempco   # ~ -2 mV/K


def test_the_closed_form_inverse_agrees_with_the_simulation():
    for i in (1e-3, 1.0, 50.0):
        assert p.shockley_voltage_for_current(i) == pytest.approx(
            _v_forward(i), rel=1e-4)


@pytest.mark.parametrize("kw,frag", [
    ({"I_S": 0.0}, "I_S"),
    ({"n": 0.0}, "emission"),
    ({"V_T": -1.0}, "V_T"),
    ({"BV": -5.1}, "magnitude"),
])
def test_unphysical_parameters_are_refused_by_name(kw, frag):
    b = p.CircuitBuilder()
    with pytest.raises(Exception, match=frag):
        b.add_shockley_diode("D1", "a", "gnd", **kw)


def test_it_beats_a_fixed_drop_model_where_that_matters():
    """The motivating comparison, end to end. A fixed-V_F model
    fitted at one current is wrong in opposite directions at the
    two ends of a converter's load range."""
    fixed_vf = 0.7
    light, heavy = 1e-3, 10.0
    v_light, v_heavy = _v_forward(light), _v_forward(heavy)
    assert v_light < fixed_vf < v_heavy
    # And the spread is large enough to matter for loss: the
    # fixed model overstates light-load drop by >20% and
    # understates full-load drop.
    assert (fixed_vf - v_light) / v_light > 0.2
    assert v_heavy > fixed_vf
